lib/chitter.rb

```diff
@@ -0,0 +1,94 @@
+require 'sinatra/base'
+require 'data_mapper'
+require 'rack-flash'
+
+# ----- DATAMAPPER SETUP BEGIN --------#
+
+env = ENV['RACK_ENV'] || 'development'
+
+DataMapper.setup(:default, "postgres://localhost/chitter_#{env}")
+
+require './lib/user'
+require './lib/peep'
+
+DataMapper.finalize
+
+DataMapper.auto_upgrade!
+
+# ------ DATAMAPPER SETUP END ---------#
+
+class Chitter < Sinatra::Base
+
+  enable :sessions
+  set :session_secret, 'seekrit' # why does this have to be set?
+
+  use Rack::Flash
+  use Rack::MethodOverride
+
+  get '/' do
+    @peeps = Peep.all
+    @peeps.reverse!
+
+    erb :index
+  end
+
+  post '/' do
+    current_user.peep.create(message: params[:peep_message], timestamp: Time.now)
+    redirect to('/')
+    erb :index
+  end
+
+  get '/users/new' do
+    erb :'/users/new'
+  end
+
+  post '/users' do
+    @user = User.new(name: params[:name],
+                     username: params[:username],
+                     email: params[:email],
+                     password: params[:password])
+    if @user.save
+      session[:user_id] = @user.id
+      redirect to('/')
+    else
+      flash.now[:errors] = @user.errors.full_messages
+      erb :'/users/new'
+    end
+  end
+
+  get '/sessions/new' do
+    erb :'/sessions/new'
+  end
+
+  post '/sessions' do
+    email = params[:email]
+    password = params[:password]
+    user = User.authenticate(email, password)
+
+    if user
+      session[:user_id] = user.id
+      redirect to('/')
+    else
+      flash[:notice] = "Sorry, login invalid"
```

Comment:
Be careful to use the correct `flash` implementation.  Should this be `flash.now`?  Do you know the difference?